README.md

```diff
@@ -86,22 +86,45 @@ Here is an example output:
 ```console
 $ sonar ps
 
-2023-07-27T18:29:17+02:00,somehost,8,user,,alacritty,3.7,214932
-2023-07-27T18:29:17+02:00,somehost,8,user,,slack,2.4,1328412
-2023-07-27T18:29:17+02:00,somehost,8,user,,X,0.8,173148
-2023-07-27T18:29:17+02:00,somehost,8,user,,brave,15.5,7085968
-2023-07-27T18:29:17+02:00,somehost,8,user,,.zoom,37.8,1722564
+v=0.7.0,time=2023-07-29T17:45:37+02:00,host=somehost,cores=12,user=someone,job=0,cmd=.vim-wrapped,cpu%=1.9,cpukib=7228,gpus={},gpu%=0,gpumem%=0,gpukib=0
+v=0.7.0,time=2023-07-29T17:45:37+02:00,host=somehost,cores=12,user=someone,job=0,cmd=node,cpu%=1.8,cpukib=79332,gpus={},gpu%=0,gpumem%=0,gpukib=0
+v=0.7.0,time=2023-07-29T17:45:37+02:00,host=somehost,cores=12,user=someone,job=0,cmd=slack,cpu%=0.7,cpukib=591720,gpus={},gpu%=0,gpumem%=0,gpukib=0
+v=0.7.0,time=2023-07-29T17:45:37+02:00,host=somehost,cores=12,user=someone,job=0,cmd=X,cpu%=1.5,cpukib=224416,gpus={},gpu%=0,gpumem%=0,gpukib=0
+v=0.7.0,time=2023-07-29T17:45:37+02:00,host=somehost,cores=12,user=someone,job=0,cmd=brave,cpu%=12.1,cpukib=3075300,gpus={},gpu%=0,gpumem%=0,gpukib=0
+v=0.7.0,time=2023-07-29T17:45:37+02:00,host=somehost,cores=12,user=someone,job=0,cmd=alacritty,cpu%=1.2,cpukib=286196,gpus={},gpu%=0,gpumem%=0,gpukib=0
+v=0.7.0,time=2023-07-29T17:45:37+02:00,host=somehost,cores=12,user=someone,job=0,cmd=sonar,cpu%=9,cpukib=372,gpus={},gpu%=0,gpumem%=0,gpukib=0
 ```
 
 The columns are:
-- time stamp
-- hostname
-- number of cores on this node
-- user
-- Slurm job ID (empty if not applicable)
-- process
-- CPU percentage (as they come out of `ps`)
-- memory used in KiB
+- `v`: version (to make data migrations easier in future)
+- `time`: local time stamp
+- `host`: host name
+- `cores`: number of cores on this node
+- `user` : username owning the process/command
+- `job`: job ID (0 if not applicable)
+- `cmd`: process/command
+- `cpu%`: CPU percentage (as they come out of `ps`)
+- `cpukib`: CPU memory used in KiB
+- `gpus`: GPU devices (more about it below)
+- `gpu%`: GPU percentage
+- `gpumem%`: GPU memory percentage
+- `gpukib`: GPU memory used in KiB
+
+`gpumem%` vs `gpumem%`:
+The difference is that on some cards some of the time it is possible to
+determine one of these but not the other, and vice versa. For example, on the
+NVIDIA cards we have we can read both quantities for running processes but only
+`gpukib` for some zombies. Since we can detect the total amount of memory here
+we could translate `gpukib` into `gpumem%`, though. On the other hand, on our
+AMD cards there is no support for detecting the absolute amount of memory used,
+nor the total amount of memory on the cards, only the percentage of gpu memory
+used. Rather than encoding the logic for dealing with this, it seemed better
+for the time being to report what we can report and let the analyzer sort it
+out.
+
+`gpus` are GPU devices:
+- If a process would use GPUs 1, 3, and 7: `"gpus={1, 3, 7}"`
+- If a process would use no or unknown GPUs: `gpus={}`
```

Comment:
There's a difference between "no gpus" and "unknown gpus".  The reason for this is that in the code for nvidia cards, most information is grabbed using `nvidia-smi pmon`, and this gives us precise information, but it does not report everything, in particular, it does not report defunct or zombie processes that hold onto GPU memory.  For that we use `nvidia-smi --query-compute-apps`, but that command does not give us information about the cards being used, and so if there are any of those we use the "unknown" tag.

While it would be nice to clean that up by using the programmatic interface to the nvidia cards (and I really should do that...), I think it is useful to preserve the distinction between "no gpus" and "unknown gpus" here, it allows us to express more.

I don't understand why the braces are desirable in the syntax here.  From a parsing point of view, the field value "gpus=1,3,7" seems almost ideal because it's just a string split operation in the ingestor.  And with that background, we could use eg "gpus=none" for the "no gpu" case (though see below) and "gpus=unknown" for the unknown case, it's easy to distinguish those two cases from the list-of-card-numbers case.

On that topic, a useful optimization for the log writer is the use of default values that are agreed between the logger and the ingestor.  The default value for `gpus` would be `none` and sensibly the field does not appear in the output at all.  For several other fields, a sensible default is zero and if the value is zero we could (optionally of course) omit the field.  The ingestor will have to deal with missing fields and obsoleted fields and newly-introduced fields anyway, so this adds very little complexity.

Reply:
Yeah... if the fields are named they can be in any order, though, and my assumption with having named fields is that fields can be introduced and removed when needed.  And indeed a log file for a given host and day can have records with and without a particular field for that reason.  The ingestion code I have in the sonarlog library handles that properly (and by design).  I think we're going to be adding quite a number of fields over time and having defaults would be nice.

The alternative is running some migration pass every time we upgrade sonar.  This sounds unpleasant - pause all cronjobs on all nodes, upgrade all log files, upgrade sonar, restart cronjobs - and error prone.  But is this what you had in mind?

Reply:
I now also think that unordered named comma-separated fields is better than forcing it into a tabular format just for the sake of file suffix which in the example I should perhaps change away from `.csv`.

The version is there just in case somebody in future decides to migrate some data but I don't plan that on any regular basis. This is just to make it possible and us less worried about breaking format.

Reply:
The high bit is that we agree that the *syntax* is CSV with comma-separation and rules for quoting, so that we can use existing high-quality CSV parsers, even if the particular *semantic* detail of the tabular format is ignored.  Both Rust and Go have such parsers that appear to be flexible about the tabular format.

Reply:
To close the loop on the gpu fields: they are samples (or close enough) and fine as it is.  The only problems we have are with cpu%, as discussed at length on bug #68.